### PR TITLE
[SPARK-17142][SQL] Complex query triggers binding error in HashAggregateExec

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

In `ReorderAssociativeOperator` rule, we extract foldable expressions with Add/Multiply arithmetics, and replace with eval literal. For example, `(a + 1) + (b + 2)` is optimized to `(a + b + 3)` by this rule.
For aggregate operator, output expressions should be derived from groupingExpressions, current implemenation of `ReorderAssociativeOperator` rule may break this promise. A instance could be:

```
SELECT
  ((t1.a + 1) + (t2.a + 2)) AS out_col
FROM
  testdata2 AS t1
INNER JOIN
  testdata2 AS t2
ON
  (t1.a = t2.a)
GROUP BY (t1.a + 1), (t2.a + 2)
```

`((t1.a + 1) + (t2.a + 2))` is optimized to `(t1.a + t2.a + 3)`, which could not be derived from `ExpressionSet((t1.a +1), (t2.a + 2))`.
Maybe we should improve the rule of `ReorderAssociativeOperator` by adding a GroupingExpressionSet to keep Aggregate.groupingExpressions, and respect these expressions during the optimize stage.
## How was this patch tested?

Add new test case in `ReorderAssociativeOperatorSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
Cool! Thanks for picking this up. cc @JoshRosen

Could you make the description a bit more readable by using markdown formatting (mostly adding 1 or 3 backticks here and there)?

**Comment by hvanhovell** [comment]
I will try to get to this one ASAP.

**Comment by jiangxb1987** [comment]
@hvanhovell I've updated the description following your advice, thank you for your time!

**Comment by jiangxb1987** [comment]
ping @hvanhovell

**Comment by jiangxb1987** [comment]
@hvanhovell Could you review this PR when you have some time please? Thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -57,20 +57,37 @@ object ConstantFolding extends Rule[LogicalPlan] {
...
-    case Multiply(l, r) => flattenMultiply(l) ++ flattenMultiply(r)
+  private def flattenMultiply(
+    expression: Expression,
+    groupSet: ExpressionSet): Seq[Expression] = expression match {
+    case expr @ Multiply(l, r) if !groupSet.contains(expr) =>
+      flattenMultiply(l, groupSet) ++ flattenMultiply(r, groupSet)
     case other => other :: Nil
   }
 
+  private def collectGroupingExpressions(plan: LogicalPlan): ExpressionSet = plan match {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Lets move this into the `apply` (you already have the relevant comment there).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -57,20 +57,37 @@ object ConstantFolding extends Rule[LogicalPlan] {
...
+    case Aggregate(groupingExpressions, aggregateExpressions, child) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: it is shorter to write: `case  Aggregate(groupingExpressions, _, _) => ExpressionSet(groupingExpressions)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -57,20 +57,37 @@ object ConstantFolding extends Rule[LogicalPlan] {
...
+      ExpressionSet.apply(groupingExpressions)
+    case _ => ExpressionSet(Seq())
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: We don't need the `Seq()`.

**Comment by hvanhovell** [comment]
LGTM. Merging to master. This is a good find. Thanks!

**Comment by hvanhovell** [comment]
We should actually make sure that `Aggregate` is not resolved in these cases.

**Comment by hvanhovell** [comment]
@jiangxb1987 could also open a backport against 2.0? Thanks!

**Comment by hvanhovell** [comment]
NVM.

**Comment by jiangxb1987** [comment]
@hvanhovell Thank you! This bug was imported in spark-2.1.0, and in spark-2.0 we don't have the problem.

**Comment by jiangxb1987** [comment]
@hvanhovell Do you mean we should check other optimize rules to ensure that `Aggregate` operator is not resolved in these cases?

**Comment by hvanhovell** [comment]
No, I am saying that it might be a good idea to incorporate this rule into `Aggregate.resolved`. This might make that method to complex.

**Comment by jiangxb1987** [comment]
Sure!Will do it soon!

**Comment by jiangxb1987** [comment]
@hvanhovell In `CheckAnalysis` we have ensured all aggregateExpressions can be derived from groupingExpressions (refer to function `checkValidAggregateExpression`), but in optimize stage the rule `ReorderAssociativeOperator` broke this. I found the `Aggregate.resolved` is only checked in analyze stage so we failed to detect this case. So, maybe incorporate this rule into `Aggregate.resolved` brings little help to avoid similar problem. Should we check analysis again after we have performed all optimize rules? Thank you!

**Pull request closed without merging**